subs/pantry/src/Pantry/Storage.hs

```diff
@@ -729,6 +909,88 @@ sinkHackagePackageNames predicate sink = do
       cnt <- count [HackageCabalName ==. nameid]
       pure $ cnt > 0
 
+-- | Get the filename for the cabal file in the given directory.
+--
+-- If no .cabal file is present, or more than one is present, an exception is
+-- thrown via 'throwM'.
+--
+-- If the directory contains a file named package.yaml, hpack is used to
+-- generate a .cabal file from it.
+findOrGenerateCabalFile
+    :: forall env. (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => Path Abs Dir -- ^ package directory
+    -> RIO env (P.PackageName, Path Abs File)
+findOrGenerateCabalFile pkgDir = do
+    hpack pkgDir
+    files <- filter (flip hasExtension "cabal" . toFilePath) . snd
+         <$> listDir pkgDir
+    -- If there are multiple files, ignore files that start with
+    -- ".". On unixlike environments these are hidden, and this
+    -- character is not valid in package names. The main goal is
+    -- to ignore emacs lock files - see
+    -- https://github.com/commercialhaskell/stack/issues/1897.
```

Comment:
Nit: What if there is an emacs lock file for a cabal file that has been deleted in the meantime?
I just checked the comment above again and it seems inaccurate in that situation since you _always_ ignore files that start with a dot.

Reply:
In that case we throw the exception that there is no cabal file found!

I don't think we want to use emacs lock file for a cabal file when the original file has been intentionally deleted.